src/coreclr/src/vm/ilmarshalers.cpp

```diff
@@ -139,11 +139,17 @@ void ILDelegateMarshaler::EmitConvertContentsNativeToCLR(ILCodeStream* pslILEmit
     EmitLoadNativeValue(pslILEmit);
     pslILEmit->EmitLDTOKEN(pslILEmit->GetToken(m_pargs->m_pMT));
     pslILEmit->EmitCALL(METHOD__TYPE__GET_TYPE_FROM_HANDLE, 1, 1); // Type System.Type.GetTypeFromHandle(RuntimeTypeHandle handle)
-    pslILEmit->EmitCALL(METHOD__MARSHAL__GET_DELEGATE_FOR_FUNCTION_POINTER, 2, 1); // Delegate System.Marshal.GetDelegateForFunctionPointer(IntPtr p, Type t)
-    EmitStoreManagedValue(pslILEmit);
 
+    // COMPAT: There is a subtle difference between argument and field marshaling with Delegate types.
+    // During field marshaling, the plain Delegate type can be used even though that type doesn't
+    // represent a concrete type. Argument marshaling doesn't permit this so we use the public
+    // API which will validate that Delegate and MulticastDelegate aren't directly used.
     if (IsFieldMarshal(m_dwMarshalFlags))
     {
+        // Delegate System.Marshal.GetDelegateForFunctionPointerInternal(IntPtr p, Type t)
+        pslILEmit->EmitCALL(METHOD__MARSHAL__GET_DELEGATE_FOR_FUNCTION_POINTER_INTERNAL, 2, 1);
+        EmitStoreManagedValue(pslILEmit);
+
         // Field marshalling of delegates supports marshalling back null from native code.
         // Parameter and return value marshalling does not.
```

Comment:
Is this comment correct? As far as I can tell, null is handled for both cases.

Reply:
Agreed. I can't see to find where null is a failure. @jkoritzinsky Any insight here? I am inclined to update this comment.

Reply:
Based on my local validation of `extern static Delegate Test2(out Delegate f);`, both the return and `f` can be `null` on return.

Reply:
Yes, this comment is correct. It describes the block below. The behavior is a little weird.

Given a P/Invoke signature: `Foo (ref Action a)`. The native signature would be (for simplicity) `(void** ptr)`. If the native code sets `*ptr = nullptr`, we don't update `a` when unmarshaling.

However, we do unmarshal `nullptr` back to `null` for delegate fields.

Reply:
>If the native code sets *ptr = nullptr, we don't update a when unmarshaling.

Ah so the behavior here is if the native call sets `null` there are no visible side-effects in managed? Which would make the below true, which seems odd...

```CSharp
extern static Delegate ReturnNull();
...

Delegate foo = <FooImpl>;
foo = ReturnNull();
Assert(foo == <FooImpl>);
```

Reply:
That's because you added a call to `EmitStoreManagedValue(pslILEmit)`.

Reply:
Sorry, that was wrong. The actual check is the `BRFALSE` to `pNullLabel` emitted above that drives the behavior difference.

Reply:
>That's because you added a call to EmitStoreManagedValue(pslILEmit).

Nope that was always there.

>Sorry, that was wrong. The actual check is the BRFALSE to pNullLabel emitted above that drives the behavior difference.

Ah. I also determined this is related to `ref` vs `out` semantics. As mentioned `ref` doesn't update, but `out` does. I will clarify the comments here.

Reply:
See https://github.com/dotnet/runtime/pull/34655